sourcegraph/cody
# Autocomplete: Fix overeagerness in returning in-flight completions

## Description

This PR fixes an issue the the `smart-throttle` experiment, where in-flight requests were incorrectly marked as `suggested` completion events. Reducing CAR and affecting other metrics.

Closes https://linear.app/sourcegraph/issue/CODY-2839/[autocomplete-latency]-fix-overeagerness-in-returning-in-flight

## Problem
The `position` and `context` provided to `provideInlineCompletionItems` is not guaranteed to be accurate for the entire duration of the function. This can happen if a user types after the function is called, their position in the line will change, and the prefix we use to build the completion is not guaranteed to be accurate. 

Example: (`export` in the prefix at the time of the request, `export const` in the actual prefix after the request has resolved, incoming completion is `export function`. We need to use the actual prefix to determine that this completion will **not** be visible)

We likely didn't run into this before as we would always cancel any in-flight requests when a new `provideInlineCompletionItems` was called. However, with the `smart-throttle` we want to keep these in-flight requests.  

## Solution

Whenever we call `provideInlineCompletionItems`, we update `latestCompletionRequest`. We can use this to determine if the position has changed since the function was invoked, and update values accordingly

## Test plan

1. Enable smart-throttle
2. Open the autocomplete trace view
3. Check that the suggested/accepted completion counts are as expected

Unit tests
